9.0 fix dispatched products array bug

atm-florianm opened the pull request:
In `htdocs/fourn/commande/dispatch.php`, there is a query to get the quantity of dispatched products for each product ID in the order. In 2015 the fetch method for the query results was switched from `fetch_rows()` to `fetch_object()`, but there was a mistake in the re-mapping of column indices to column names.

atm-florianm committed:
FIX: calculation of $products_dispatched
After the DB retrieval method was changed from fetch_row to fetch_object, the indices used became wrong.

atm-florianm committed:
FIX: wrong display (and hidden input) for already dispatched quantity

eldy commented:
The use if rowid was done on purpose.
When we have 2 lines with same product, one with 10 P and one with 20 P, when we make the dispatch, we must still have the 2 lines and we must keep difference of what was shipped for line 1 and 2.
It seems your fix will break this requirement, isn't it ?

eldy commented:
The use if rowid was done on purpose.
When we have 2 lines with same product, one with 10 P and one with 20 P, when we make the dispatch, we must still have the 2 lines and we must keep difference of what was shipped for line 1 and 2.
It seems your fix will break this requirement, isn't it ?

atm-florianm commented:
> It seems your fix will break this requirement, isn't it ?

Thank you for your reply, I now understand the point of this new behavior. 

The problem we are currently facing happens when a client upgrades to 9.0 from a version without `fk_commandefourndet` in `llx_commande_fournisseur_dispatch`. The database migration scripts populate the column with 0 (default). Due to the new behavior, all past supplier orders look like the products were never dispatched. However, all supplier orders dispatched after the upgrade will be shown as dispatched, as intended.

A solution could be to enable (maybe with a hidden conf switch) a fallback to the old `$remaintodispatch` calculation method (i.e. use fk_product as a key) when the query returns rows with l.rowid == null and to keep the current behavior as default. That way, old supplier orders that were already fully dispatched would be correctly locked for dispatch but newer orders would use the new behavior. I am conscious this solves an edge case and may not be the ideal solution.

atm-florianm commented:
Another idea suggested by a colleague: I could make a migration script that would replace null `fk_commandefourndet` with the ID of the first matching order line (in terms of `fk_product`). This would make the past commits of this PR unnecessary and still enable our client to clearly see which supplier orders were fully dispatched. I could maybe put the script in a new commit in this PR to keep track of the discussion.

atm-florianm committed:
New script that links "orphaned" dispatch lines to order lines
(the "orphans" are dispatch lines that were created in old versions of Dolibarr which didn’t have fk_commandefourndet in llx_commande_fournisseur_dispatch)

atm-florianm committed:
Revert "FIX: wrong display (and hidden input) for already dispatched quantity"

This reverts commit b155fdc014caaeea15848605ae75e82823819c1f.

atm-florianm committed:
Revert "FIX: calculation of $products_dispatched"

This reverts commit 75ae611f9e49fead3aa825249b55734a5088cab4.

atm-florianm committed:
Fix problems detected by travis-ci

eldy commented:
Yes, if trouble is introduced by a bad or not complete migration, best correction is to introduce a better migration step. Using a hidden option to "keep a bugged behaviour" is rarely a good idea.

atm-florianm committed:
Removed commented out code from the new migration script

eldy commented on htdocs/fourn/commande/script/repair_llx_commande_fournisseur_dispatch_3.6-9.0.php:
```diff
@@ -0,0 +1,160 @@
+<?php
```
Adding maintenance script at same level than application pages is not a good idea.
Can you report this into the dedicated script    /install/repair.php  that is already protected against bad use ?

atm-florianm replied:
Thank you, I have now moved the code to `install/repair.php`.

atm-florianm committed:
Moved repair payload from a specific script to install/repair.php
